### PR TITLE
Logging using after_request method & hydrus_logger added

### LINKED ISSUE #572: Adding detailed logging to hydrus
### I'm submitting a 
- [ ] bug report.
- [x] feature request.

### Current Behaviour:

Currently, we don't have any detailed logging system for hydrus. No logs are being generated.
### Expected Behaviour:

We can generate the log files which could help the users for getting the `Object_id` for any future manipulation 
and developers for debugging faster.

We will have a separate log folder that will be generated while running hydrus server.

```
Logs
|
|_ _ hydrus.log
```
Log file format

```
Logs/hydrus.log

run_time : 12-may-12:00:00
API_Doc fetched Successfully.
Context generated.
Terminated!
Database setup Error.
```
### Do you want to work on this issue?

yes, I and @farazkhanfk7 are working on this and changing the API response structure [issue](https://github.com/HTTP-APIs/hydrus/issues/573) .

### PR DESCRIPTION
**Pull request by Purvanshsingh**
Fixes #572

### Checklist
- [x] My branch is up-to-date with upstream/develop branch.
- [x] Everything works and tested for Python 3.6.0 and above.

### Current behaviour

We don't have any logging mechanism for hydrus.
### New expected behaviour

I have tried to add the things covered in [this discussion](https://github.com/HTTP-APIs/hydrus/issues/572), Opening this PR for more clear Suggestions.

1. I have added a `after_request` method where we are logging the `TimeStamp job_Id Method / iri` (resource_url is not fetched now) 
2.  Created `hydrus_logger.py` which provide us configured structlog logger object
3. `LogConfig` file for configuring the logger object.
4.  Logs Directory is not setup yet, `hydrus.log` will be generated in the same directory.

Screenshot from 2021-05-15 05-27-47

### CODE REVIEW

**Comment by farazkhanfk7** [comment]
Works fine. Except for one issue that this was working fine when `fileConfig` was being used instead of `dictConfig`. Now it's not creating `hydrus.log` file after the latest commit. Please cross-check and let me know if I made any mistake here. I think we have to use relative path of LogConfig in HydrusLogger class. 

Also I can see, that `resource_url` is not being fetched yet and is hardcoded to `<iri>`. 
Suggestion : 
- You can use `request.path` to get the complete URL of path.
- Also to get the IRI of newly created/updated object you can add conditional checks for `response.status_code` and return IRI of object using `response.location`.

I trying adding these on my local. This was the format after creation of a new object:
```
2021-05-15 11:52.10 [info     ] 327bbe6e-0280-4eaf-99a3-0222e52fc170 GET:/serverapi/MovieCollection, Status Code : [200 OK]
2021-05-15 11:53.23 [info     ] 29dcabbe-1c46-48b1-bedc-b9913c24afdb PUT:/serverapi/Movie/,  Object created with ID:http://localhost:8080/serverapi/Movie/0134ee40-a090-484f-97ee-df41d993e775
```

File: hydrus/LogConfig.py
Comment (Mec-iS) [inline hydrus/LogConfig.py]: name of modules is lowercase and use underscores. please be attentive to previous decisions and maintain coherence in the codebase

File: hydrus/app_factory.py
Comment (Mec-iS) [inline hydrus/app_factory.py]: definition of logger should be done at module-level, not inside a function.

This line should be at the top of the module, after imports.

**Review by Mec-iS — CHANGES_REQUESTED** [review]
as mentioned in a comment, the log level cannot be hardcoded but needs to be set according to an environment variable, please be more attentive when reading suggestions.

`level` should get the value from Python `environ` or default to `DEBUG` if the value is null.

File: hydrus/LogConfig.py
Comment (Purvanshsingh) [reply hydrus/LogConfig.py]: This file is removed. we are not using `filedictConfig()` instead we are using `fileConfig('LogConfig')`

File: hydrus/app_factory.py
Comment (Purvanshsingh) [reply hydrus/app_factory.py:26]: > as mentioned in a comment, the log level cannot be hardcoded but needs to be set according to an environment variable, please be more attentive when reading suggestions.
> 
> `level` should get the value from Python `environ` or default to `DEBUG` if the value is null.

I just made it hardcoded for testing and verification purpose.

**Comment by Purvanshsingh** [comment]
> as mentioned in a comment, the log level cannot be hardcoded but needs to be set according to an environment variable, please be more attentive when reading suggestions.
> 
> `level` should get the value from Python `environ` or default to `DEBUG` if the value is null.

I just made it hardcoded for testing & Verification purpose.

File: hydrus/LogConfig.py
Comment (Purvanshsingh) [reply hydrus/LogConfig.py]: We have `LogConfig` file for configuring the logging object.

**Comment by Purvanshsingh** [comment]
@Mec-iS  Can we set the default logging level to `INFO` instead of `DEBUG` as we are using environment variable all the logging messages are being logged to the file.

Especially` webscoket.server` logging messages.
Screenshot from 2021-05-15 15-45-57

File: hydrus/app_factory.py
Comment (farazkhanfk7) [inline hydrus/app_factory.py]: I think this line should be added inside an if-else block ( `if response.location : `) to check if we're getting a `response.location` or not. Otherwise `Object_created_at :` will keep returning `None` even in case of GET requests to other endpoints. In that case, we should only log the method and path of request.

**Comment by Mec-iS** [comment]
> @Mec-iS Can we set the default logging level to `INFO` instead of `DEBUG` as we are using environment variable all the logging messages are being logged to the file.
> 
> Especially` webscoket.server` logging messages.

those are relevant logs, there should be a way to wrap those logs in the structlog context as well.

In general, the project is still in debugging phase so the right level is `DEBUG`. Once we move closer to a production-ready system it will be more correct to set it to `INFO`.

File: hydrus/app_factory.py
Comment (Purvanshsingh) [reply hydrus/app_factory.py]: @farazkhanfk7 I have done that, just waiting for the level `DEBUG` approval to push all the changes.
Thanks.

Comment (Mec-iS) [reply hydrus/app_factory.py]: please see my comment below

Comment (Purvanshsingh) [reply hydrus/app_factory.py]: @Mec-iS yes, I am figuring out the way to deal with those logs with structlog.

**Comment by Purvanshsingh** [comment]
> > @Mec-iS Can we set the default logging level to `INFO` instead of `DEBUG` as we are using environment variable all the logging messages are being logged to the file.
> > Especially` webscoket.server` logging messages.
> 
> those are relevant logs, there should be a way to wrap those logs in the structlog context as well.
> 
> In general, the project is still in debugging phase so the right level is `DEBUG`. Once we move closer to a production-ready system it will be more correct to set it to `INFO`.

@Mec-iS I went through the structlog context, we can only structure those messages to any format that we have logged.
I tried doing that as well but these messages are from ` webscoket.server` they can't be structured. they are being logged because we are using `DEBUG` level they won't be logged as we change to `INFO`.

**Comment by Mec-iS** [comment]
leave them there for now

**Review by sameshl — APPROVED** [review]
Looks good. Great work! :+1:

**Review by Mec-iS — APPROVED** [review]

**Pull request merged**
